crates/c-api/src/wasi.rs

```diff
@@ -270,3 +327,89 @@ pub unsafe extern "C" fn wasi_config_preopen_dir(
 
     true
 }
+
+#[no_mangle]
+pub unsafe extern "C" fn wasi_pipe_new(
+    limit: usize,
+    ret_read_pipe: *mut *mut wasi_read_pipe_t,
+    ret_write_pipe: *mut *mut wasi_write_pipe_t,
+) {
+    let queue = BoundedVecDeque::new(limit);
+    let queue = Arc::new(RwLock::new(queue));
+
+    if !ret_read_pipe.is_null() {
+        *ret_read_pipe = Box::into_raw(Box::new(wasi_read_pipe_t {
+            queue: queue.clone(),
+        }));
+    }
+    if !ret_write_pipe.is_null() {
+        *ret_write_pipe = Box::into_raw(Box::new(wasi_write_pipe_t {
+            queue: queue.clone(),
+        }));
+    }
+}
+
+#[no_mangle]
+pub extern "C" fn wasi_read_pipe_len(read_pipe: &wasi_read_pipe_t) -> usize {
+    let queue = read_pipe.queue.read().unwrap();
+    queue.len()
+}
+
+#[no_mangle]
+pub unsafe extern "C" fn wasi_read_pipe_read(
+    read_pipe: &mut wasi_read_pipe_t,
+    buf: *mut u8,
+    buf_len: usize,
+) -> usize {
+    let mut buf = crate::slice_from_raw_parts_mut(buf, buf_len);
+    let mut queue = read_pipe.queue.write().unwrap();
+    std::io::Read::read(&mut *queue, &mut buf).unwrap()
+}
+
+#[no_mangle]
+pub unsafe extern "C" fn wasi_write_pipe_write(
+    write_pipe: &mut wasi_write_pipe_t,
+    buf: *const u8,
+    buf_len: usize,
+) -> usize {
+    let buf = crate::slice_from_raw_parts(buf, buf_len);
+    let mut queue = write_pipe.queue.write().unwrap();
+    std::io::Write::write(&mut *queue, buf).unwrap()
+}
+
+#[derive(Default)]
+pub struct BoundedVecDeque<T> {
+    deque: VecDeque<T>,
+    limit: usize,
+}
+impl<T> BoundedVecDeque<T> {
+    fn new(limit: usize) -> Self {
+        Self {
+            deque: VecDeque::new(),
+            limit,
+        }
+    }
+    #[inline]
+    fn len(&self) -> usize {
+        self.deque.len()
+    }
+}
+impl io::Read for BoundedVecDeque<u8> {
+    #[inline]
+    fn read(&mut self, buf: &mut [u8]) -> io::Result<usize> {
+        io::Read::read(&mut self.deque, buf)
+    }
+}
+impl io::Write for BoundedVecDeque<u8> {
+    #[inline]
+    fn write(&mut self, buf: &[u8]) -> io::Result<usize> {
+        let amt = self.limit.saturating_sub(self.len());
+        let amt = std::cmp::min(amt, buf.len());
+        self.deque.extend(&buf[..amt]);
+        Ok(amt)
+    }
```

Comment:
I'm not sure that this read/write impl are the best for WASI. As modeled today I think it would be best to implement synchronous I/O here but this is sort of a form of nonblocking I/O without the ability to be notified  when bytes are ready.

Specifically `Read` will not block when the deque is empty, instead returning `Ok(0)` which many libraries actually interpret as EOF. Additionally `Write` will not block when it's full, instead returning `Ok(0)` which I think libraries take various viewpoints on.

I think it would be better for `Read` to block on writing and for writing to block on `Read`.

Reply:
UNIX pipe returning 0 bytes read means not more data **for now**

Synchronous IO seems be more user friendly.
But there if not way to read all of stdin for example.
It would just be a deadlock after the last byte.

Furthermore I don't see how we can't contact the runtime to suspend the execution and run the "writer" instance.

Reply:
I don't think unix pipes return 0 when there's no more data, when the write pipe is still open then it blocks waiting for more data. I think the same thing should happen here, the reader blocks until the writer has written data or the writer has been dropped.